File: src/main/java/corp/soyummy/auth/AuthService.java
package corp.soyummy.auth;

import org.springframework.stereotype.Service;
import lombok.AllArgsConstructor;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import corp.soyummy.config.JwtService;

import corp.soyummy.auth.dto.LoginRequestDTO;
import corp.soyummy.auth.dto.RegisterRequestDTO;

@Service
@AllArgsConstructor
public class AuthService {
    private final AuthRepository authRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    public String register(RegisterRequestDTO request, Role role) {
        User user = User.builder()
                .name(request.getName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(role)
                .build();

        String token = jwtService.generateToken(user.getEmail());
        user.setToken(token);

        authRepository.save(user);

        return token;
    }

    public String login(LoginRequestDTO request) {
        User user = authRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new UsernameNotFoundException("User not found"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new BadCredentialsException("Invalid password");
        }

        String token = jwtService.generateToken(user.getEmail());
        user.setToken(token);
        authRepository.save(user);

        return token;
    }

    public void logout(String token) {
        if (token == null || token.isBlank()) {
            throw new IllegalArgumentException("Token is required");
        }

        if (token.startsWith("Bearer ")) {
            token = token.substring(7);
        } else {
            throw new IllegalArgumentException("Invalid token format");
        }

        User user = authRepository.findByToken(token)
                .orElseThrow(() -> new IllegalArgumentException("Invalid token"));

        user.setToken(null);

        authRepository.save(user);
    }
}

File: src/main/kotlin/com/soyummy/auth/AuthService.kt
package com.soyummy.auth

import org.springframework.stereotype.Service
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.userdetails.UsernameNotFoundException
import com.soyummy.config.JwtService
import com.soyummy.auth.dto.LoginDTO
import com.soyummy.auth.dto.RegisterDTO

@Service
class AuthService(
    private val authRepository: AuthRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtService: JwtService
) {

    fun register(request: RegisterDTO, role: Role): String {
        val user = User(
            name = request.name,
            email = request.email,
            password = passwordEncoder.encode(request.password),
            role = role
        )

        val token = jwtService.generateToken(user.email)
        user.token = token

        authRepository.save(user)

        return token
    }

    fun login(request: LoginDTO): String {
        val user = authRepository.findByEmail(request.email)
            .orElseThrow { UsernameNotFoundException("User not found") }

        if (!passwordEncoder.matches(request.password, user.password)) {
            throw BadCredentialsException("Invalid password")
        }

        val token = jwtService.generateToken(user.email)
        user.token = token
        authRepository.save(user)

        return token
    }

    fun logout(token: String) {
        if (token.isBlank()) {
            throw IllegalArgumentException("Token is required")
        }

        val cleanedToken = if (token.startsWith("Bearer ")) {
            token.substring(7)
        } else {
            throw IllegalArgumentException("Invalid token format")
        }

        val user = authRepository.findByToken(cleanedToken)
            .orElseThrow { IllegalArgumentException("Invalid token") }

        user.token = null
        authRepository.save(user)
    }
}